tmpl/wiki/package-quality.md

```diff
@@ -0,0 +1,138 @@
+  strongly encouraged but not yet mandatory as 1.0.0 has not been
+  officially released. However, we plan to enforce its use as soon as
+  it is the case -- assuming that it has all the features needed for
+  building MirageOS applications.
+
+- Use proper names: same name for opam and ocamlfind package, and also ideally
+  for the top-level module.
+
+- All direct dependencies are specified in the opam file (e.g. sometimes some
+  direct dependencies are included by transitivity -- this is not very stable
+  and could break the package in the future if one of a transitive dependency
+  is updated) (NOTE: should probably be added in the packaging guideline)
+
+### ![Level 2](https://img.shields.io/badge/level-2-blue.svg) Define Package Scope
+
+- Have well-identified maintainers. GitHub recently
+  [introduced code owners](https://github.com/blog/2392-introducing-code-owners),
+  which may be worth to use for some (of our core) libraries.
+
+- Have a design document specifying the scope of the library (and esp. what is
+  out of scope)
+
+- Use the correct set of tags in opam metadata
+
+- Have documentation
+
+### ![Level 3:](https://img.shields.io/badge/level-3-blue.svg) Use Good Coding Style
+
+- Have proper indentation (using ocp-indent + checked-in ocp-indent file).
+
+- Use at most 80 columns (`ocp-indent` unfortunately doesn't check this).
+
+- Follow [OCaml programming guidelines](https://ocaml.org/learn/tutorials/guidelines.html).
+
+### ![Level 4:](https://img.shields.io/badge/level-4-blue.svg) Keep your Style Functional
+
+- Avoid global mutable state.
+
+- Use tail recursion.
+
+- Use `-safe-string` (looks like this will be the default in 4.06).
+
+- Avoid polymorphic equality and comparison -- when possible define your own
+  specialized functions.
+
+### ![Level 5:](https://img.shields.io/badge/level-5-blue.svg) Test
+
+- Have unit tests (using alcotest) (ideally with coverage report).
+
+### ![Level 6:](https://img.shields.io/badge/level-6-blue.svg) Keep Sane Dependencies
+
+- Depends only on libraries released in opam.
+
+- Provide pretty printers by using [fmt](http://erratique.ch/software/fmt/doc/Fmt.html).
+
+- Do not expose any exceptions in the public interface, but use the
+  [result](http://caml.inria.fr/pub/docs/manual-ocaml/libref/Pervasives.html#TYPEresult)
+  type and [rresult](http://erratique.ch/software/rresult/doc/Rresult.html)
+  combinators.
+
+- Use [logs](http://erratique.ch/software/logs/doc/Logs.html) for logging.
+
+- Use [astring](http://erratique.ch/software/astring/doc/Astring.html) for
+  string processing.
+
+- Use [bos](http://erratique.ch/software/bos/doc/Bos.html) for operating system
+  interaction.
+
+- Use [fpath](http://erratique.ch/software/fpath/doc/Fpath.html) for file paths.
+
+- Avoid the use of C bindings for no good reasons. A good reason would be to
+  improve performance by an order of magnitude, or re-use an existing C library
+  that has not been rewritten yet.
```

Comment:
maybe elaborate a bit with links how difficult it is to get C bindings cross compiled atm!?

Reply:
We definitely should add a documentation. What I'm afraid with MirageOS 4 is the ability to compile C stubs for the Unix target (with the `libc` according to the `dune` context) and be blocked when we want to switch the whole codebase to a new target such as `ocaml-freestanding`/`Solo5`.

Currently, `mirage-crypto` or `digestif` try to compile _properly_ C stubs with `ocaml-freestanding` flags (including `nolibc`) but such test (just to check that these C stubs can be _cross-compiled_ with `ocaml-freestanding`) is not done by `mirage` and nothing (as far as I can say) can prevent a bad C stubs for all targets.